### PR TITLE
Add propkey defs and most of propidl

### PR DESCRIPTION
**Pull request by austinwagner**
The entire propkey.h file is ported over. PROPVARIANT now has a union inside instead of a byte array. The union has fields for all types that currently exist in 0.3. The remainder are commented out to easily add later.

### CODE REVIEW

**Comment by austinwagner** [comment]
I decided I might as well do as much of propidl.rs as I could

**Comment by austinwagner** [comment]
Rebased

File: src/macros.rs [hunk 1]
```diff
@@ -371,3 +371,9 @@ macro_rules! FN {
         pub type $func = Option<unsafe extern "C" fn($($p: $t,)*) -> $ret>;
     );
 }
+macro_rules! DEFINE_CA {
```
Comment (retep998) [inline src/macros.rs]: This macro is named `TYPEDEF_CA` in the `.idl` file.

Comment (retep998) [reply src/macros.rs]: Also, since it is only used in one module, it should probably be defined in that module too.

File: src/um/propidl.rs [hunk 2]
```diff
@@ -4,13 +4,384 @@
...
+STRUCT!{struct STATPROPSETSTG {
+    fmtid: FMTID,
+    clsid: CLSID,
+    grfFlags: DWORD,
+    mtime: FILETIME,
+    ctime: FILETIME,
+    atime: FILETIME,
+    dwOSVersion: DWORD,
+}}
+RIDL!(#[uuid(0x00000138, 0x0000, 0x0000, 0xC0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x46)]
```
Comment (retep998) [inline src/um/propidl.rs]: `RIDL` is an item level macro and should use `{}` instead of `()`.

File: src/um/propidl.rs [hunk 3]
```diff
@@ -4,13 +4,384 @@
...
+        fReserved: BOOLEAN,
+        pcIndirect: *mut ULONG,
+    ) -> *mut SERIALIZEDPROPERTYVALUE;
+    pub fn StgConvertPropertyToVariant(
+        pprop: *const SERIALIZEDPROPERTYVALUE,
+        CodePage: USHORT,
+        pvar: *mut PROPVARIANT,
+        pma: *mut PMemoryAllocator
+    ) -> BOOLEAN;
+}
```
Comment (retep998) [inline src/um/propidl.rs]: Missing end of file newline.

File: src/um/propkey.rs [hunk 4]
```diff
@@ -0,0 +1,2309 @@
...
+    0x64440490, 0x4C8B, 0x11D1, 0x8B, 0x70, 0x08, 0x00, 0x36, 0xB1, 0x1A, 0x03, 10);
+DEFINE_PROPERTYKEY!(PKEY_Audio_EncodingBitrate,
+    0x64440490, 0x4C8B, 0x11D1, 0x8B, 0x70, 0x08, 0x00, 0x36, 0xB1, 0x1A, 0x03, 4);
+DEFINE_PROPERTYKEY!(PKEY_Audio_Format,
+    0x64440490, 0x4C8B, 0x11D1, 0x8B, 0x70, 0x08, 0x00, 0x36, 0xB1, 0x1A, 0x03, 2);
+DEFINE_PROPERTYKEY!(PKEY_Audio_IsVariableBitRate,
+    0xE6822FEE, 0x8C17, 0x4D62, 0x82, 0x3C, 0x8E, 0x9C, 0xFC, 0xBD, 0x1D, 0x5C, 100);
+DEFINE_PROPERTYKEY!(PKEY_Audio_PeakValue,
+    0x2579E5D0, 0x1116, 0x4084, 0xBD, 0x9A, 0x9B, 0x4F, 0x7C, 0xB4, 0xDF, 0x5E, 100);
+DEFINE_PROPERTYKEY!(PKEY_Audio_SampleRate,
```
Comment (retep998) [inline src/um/propkey.rs]: `DEFINE_PROPERTYKEY` is an item level macro and should use `{}` instead of `()`.

File: src/um/propkey.rs [hunk 5]
```diff
@@ -0,0 +1,2309 @@
...
+DEFINE_PROPERTYKEY!(PKEY_Devices_Aep_IsConnected,
+    0xA35996AB, 0x11CF, 0x4935, 0x8B, 0x61, 0xA6, 0x76, 0x10, 0x81, 0xEC, 0xDF, 7);
+DEFINE_PROPERTYKEY!(PKEY_Devices_Aep_IsPaired,
+    0xA35996AB, 0x11CF, 0x4935, 0x8B, 0x61, 0xA6, 0x76, 0x10, 0x81, 0xEC, 0xDF, 16);
+DEFINE_PROPERTYKEY!(PKEY_Devices_Aep_IsPresent,
+    0xA35996AB, 0x11CF, 0x4935, 0x8B, 0x61, 0xA6, 0x76, 0x10, 0x81, 0xEC, 0xDF, 9);
+DEFINE_PROPERTYKEY!(PKEY_Devices_Aep_Manufacturer,
+    0xA35996AB, 0x11CF, 0x4935, 0x8B, 0x61, 0xA6, 0x76, 0x10, 0x81, 0xEC, 0xDF, 5);
+DEFINE_PROPERTYKEY!(PKEY_Devices_Aep_ModelId,
+    0xA35996AB, 0x11CF, 0x4935, 0x8B, 0x61, 0xA6, 0x76, 0x
```
Comment (retep998) [inline src/um/propkey.rs]: Missing end of file newline.

File: src/um/propidl.rs [hunk 6]
```diff
@@ -4,13 +4,390 @@
...
+pub const PROPSETFLAG_ANSI: DWORD = 2;
+pub const PROPSETFLAG_UNBUFFERED: DWORD = 4;
+pub const PROPSET_BEHAVIOR_CASE_SENSITIVE: DWORD = 1;
+STRUCT!{struct VERSIONEDSTREAM {
+    guidVersion: GUID,
+    pStream: *mut IStream,
+}}
+pub type LPVERSIONEDSTREAM = *mut VERSIONEDSTREAM;
+macro_rules! TYPEDEF_CA {
+    ($name:ident, $t:ty) => { STRUCT!{struct $name {
```
Comment (retep998) [inline src/um/propidl.rs]: Your macro takes the parameters in the reverse order that it does in the original headers, nor do the parameter names match the original parameter names. `#define TYPEDEF_CA(type, name)`

File: src/shared/wtypes.rs [hunk 7]
```diff
@@ -87,6 +87,17 @@ pub type wireBSTR = *mut FLAGGED_WORD_BLOB;
...
+STRUCT!{struct BSTRBLOB {
+    cbSize: ULONG,
+    pData: *mut BYTE,
+}}
+pub const VARIANT_TRUE: VARIANT_BOOL = -1;
+pub const VARIANT_FALSE: VARIANT_BOOL = 0;
+STRUCT!{struct CLIPDATA {
+    cbSize: ULONG,
+    ulClipFmt: LONG,
+    pClipData: BYTE,
```
Comment (retep998) [inline src/shared/wtypes.rs]: `BYTE * pClipData;`
This should be a pointer!

**Pull request merged**